### PR TITLE
Implemented NoViewID

### PR DESCRIPTION
**Pull request by dastgir**
(http://herc.ws/board/topic/10020-official-mapflag-to-disable-viewid/)

### CODE REVIEW

File: src/map/atcommand.c [hunk 1]
```diff
@@ -3873,6 +3873,8 @@ ACMD(mapinfo) {
 		strcat(atcmd_output, msg_fd(fd,1096)); // PartyLock |
 	if (map->list[m_id].flag.guildlock)
 		strcat(atcmd_output, msg_fd(fd,1097)); // GuildLock |
+	if (map->list[m_id].flag.guildlock)
```
Comment (4144) [inline src/map/atcommand.c]: copy/paste error?

Comment (dastgir) [inline src/map/atcommand.c]: Indeed

File: src/map/map.h [hunk 2]
```diff
@@ -716,6 +718,7 @@ struct map_data {
 		unsigned noknockback : 1;
 		unsigned notomb : 1;
 		unsigned nocashshop : 1;
+		unsigned noviewid : 22;
```
Comment (AnnieRuru) [inline src/map/map.h]: where did you get the value 22 ?
is there a constant ? EQI_MAX ?

**Comment by AnnieRuru** [comment]
we don't have a doc\mapflag.txt like rathena do
https://github.com/rathena/rathena/blob/master/doc/mapflags.txt

**Comment by AnnieRuru** [comment]
I got a feeling this whole patch looks funny 
in `setmapflag` script command, we have this
`map->list[m].flag.noviewid = (val <= 0) ? 0 : val; break;`
then when checking the condition, it is
`if (!map_no_view(sd->bl.m,EQP_COSTUME_GARMENT) && .... `
means we have to know the actual bitmask of EQP_COSTUME_GARMENT, which is 0x002000
but the limit seems to be just 22, something not right

File: src/map/map.h (hunk 2 above)
Comment (dastgir) [inline src/map/map.h]: Max Value is EQP_SHADOW_ACC_L = 0x200000, i.e 22 bit,
the : 22 means 22 Bit, and not value 22

**Comment by dastgir** [comment]
22 is the Bit, and not the Maximum value

**Comment by AnnieRuru** [comment]
hahaha, come laugh at me for someone who doesn't take class in C
ok so, we should define a `EQP_MAX` in pc.h,
will it give compile error ? hmm ...

File: src/map/map.h [hunk 3]
```diff
@@ -227,6 +227,8 @@ enum {
 #define map_flag_gvg2(m) (map->list[m].flag.gvg || map->list[m].flag.gvg_castle)
 // No Kill Steal Protection
 #define map_flag_ks(m) (map->list[m].flag.town || map->list[m].flag.pvp || map->list[m].flag.gvg || map->list[m].flag.battleground)
+// No ViewID
+#define map_no_view(m, view) (map->list[m].flag.noviewid & view)
```
Comment (MishimaHaruna) [inline src/map/map.h]: You might want to add parentheses around `view`, to make this safer.

```
map_no_view(m, foo|bar);  // (map->list[m].flag.noview & foo | bar)
```

**Comment by dastgir** [comment]
Done @MishimaHaruna

**Comment by MishimaHaruna** [comment]
Merged, thank you!

**Comment by theultramage** [comment]
The '22' bits thing is really awful, please change it to a simple integer!
Might be useful to add ```// enum equip_pos bitmask``` next to the declaration in map.h.
Also, please fix the tabs-instead-of-spaces indentation used in script.c in that patch.

**Comment by MishimaHaruna** [comment]
Hmm, good points indeed.

**Pull request closed without merging**